### PR TITLE
[BrainPOP] Add new extractor (Closes #10000)

### LINKED ISSUE #10000: Site support request: Brainpop
Example URL: https://www.brainpop.com/english/freemovies/williamshakespeare/

### PR DESCRIPTION
**Pull request by nehalvpatel**
### Before submitting a _pull request_ make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
### What is the purpose of your _pull request_?
- [ ] Bug fix
- [x] New extractor
- [ ] New feature

---
### Description of your _pull request_ and other information

Added a new extractor for [BrainPOP](https://www.brainpop.com/). In addition to the video, it extracts the description and the thumbnail. It should work on any video on the site, even if it's in a different language.

### CODE REVIEW

File: youtube_dl/extractor/brainpop.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+
+        ec_token = self._html_search_regex(r'ec_token : \'(.+)\'', webpage, 'token')
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Its easier to parse the JSON instead of searching through it manually.

File: youtube_dl/extractor/brainpop.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        thumbnail_cdn = self._html_search_regex(r"'cdn_path' : '([^']*)'", webpage, 'thumbnail cdn', fatal=False)
+        thumbnail_image = self._html_search_regex(r'type":"Movie","name":"[^"]*","image":"([^"]*)"', webpage, 'thumbnail', fatal=False)
+        thumbnail = thumbnail_cdn + thumbnail_image.replace('\\', '')
+
+        description = self._html_search_regex(r'type":"Movie","name":"[^"]*","image":"[^"]*","description":"([^"]*)"', webpage, 'description', fatal=False)
+
+        return {
+            'id': video_id,
+            'url': url,
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: There are multiple formats

File: youtube_dl/extractor/brainpop.py (hunk 1 above)
Comment (nehalvpatel) [inline youtube_dl/extractor/brainpop.py]: @TRox1972 Right, but the JSON is mixed in with the HTML. How should I go about extracting it?

File: youtube_dl/extractor/brainpop.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+
+        ec_token = self._html_search_regex(r"ec_token : '([^']*)'", webpage, 'token')
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: It this needed? Does it work without it?

File: youtube_dl/extractor/brainpop.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        })
+        formats.append({
+            'url': movie_cdn_path + movies['mp4_small'] + '?' + ec_token,
+            'height': 480,
+            'width': 480,
+        })
+        self._sort_formats(formats)
+
+        thumbnails = []
+        for (i, screenshot) in enumerate(screenshots):
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Just loop through `screenshots`; you aren't using `i` anyway

Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Also you can shorten this to a block statement, I guess

File: youtube_dl/extractor/brainpop.py (hunk 3 above)
Comment (nehalvpatel) [inline youtube_dl/extractor/brainpop.py]: The token? Yea, you can't access the videos without it.

File: youtube_dl/extractor/brainpop.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        title = settings['title']
+        description = settings['description']
+
+        global_content = self._parse_json(self._html_search_regex(r'var global_content = ([^;]*)', webpage, 'global content').replace("'", '"'), video_id)
+        cdn_path = global_content['cdn_path']
+        movie_cdn_path = global_content['movie_cdn_path']
+
+        content = self._parse_json(self._html_search_regex(r'var content = ([^;]*)', webpage, 'content'), video_id)
+        movies = content['category']['unit']['topic']['movies']
+        screenshots = content['category']['unit']['topic']['screenshots']
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: This should be non-fatal

File: youtube_dl/extractor/brainpop.py (hunk 4 above)
Comment (nehalvpatel) [inline youtube_dl/extractor/brainpop.py]: Sorry, I'm new to Python. Could you show me how?

File: youtube_dl/extractor/brainpop.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+
+        ec_token = self._html_search_regex(r"ec_token : '([^']*)'", webpage, 'token')
+
+        settings = self._parse_json(self._html_search_regex(r'var settings = ([^;]*)', webpage, 'settings'), video_id)
+        title = settings['title']
+        description = settings['description']
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: It's better to return `title` and `description` directly since you are not using the variables anywhere else. Also, description extraction should be non-fatal

File: youtube_dl/extractor/brainpop.py (hunk 5 above)
Comment (nehalvpatel) [inline youtube_dl/extractor/brainpop.py]: Which part? The settings JSON?

File: youtube_dl/extractor/brainpop.py (hunk 4 above)
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]:
```
thumbnails = [{'url': cdn_path + screenshot} for screenshot in screenshots]

File: youtube_dl/extractor/brainpop.py (hunk 5 above)
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Nope, screenshots, e.g. `screenshots = content['category']['unit']['topic'].get('screenshots', {}`

File: youtube_dl/extractor/brainpop.py (hunk 6 above)
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: You can also remove the '` - BrainPOP`' string from the title by using `from ..utils import remove_end`

File: youtube_dl/extractor/brainpop.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+
+class BrainPOPIE(InfoExtractor):
+    _VALID_URL = r'https?:\/\/(?:(.+)\.)?brainpop\.com\/(?P<id>[^\r\n]+)'
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Only the last part of the URL should be the `id`

File: youtube_dl/extractor/brainpop.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            'id': 'english/freemovies/williamshakespeare/',
+            'ext': 'mp4',
+            'title': 'William Shakespeare - BrainPOP',
+            'thumbnail': 're:^https?://.*\.png$',
+            'description': 'He could do comedies, tragedies, histories and poetry.  Learn about the greatest playwright in the history of the English language!',
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Use `display_id` instead; the 'real' `ID` seems to be stored in the JSON as `EntryID`

File: youtube_dl/extractor/brainpop.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        self.report_extraction(video_id)
+
+        ec_token = self._html_search_regex(r"ec_token : '([^']*)'", webpage, 'token')
+
+        settings = self._parse_json(self._html_search_regex(r'var settings = ([^;]*)', webpage, 'settings'), video_id)
+        title = settings['title']
+        description = settings['description']
+
+        global_content = self._parse_json(self._html_search_regex(r'var global_content = ([^;]*)', webpage, 'global content').replace("'", '"'), video_id)
+        cdn_path = global_content['cdn_path']
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: This should be non-fatal

File: youtube_dl/extractor/brainpop.py [hunk 10]
```diff
@@ -50,17 +52,14 @@ def _real_extract(self, url):
...
+        settings = self._parse_json(self._html_search_regex(r'var settings = ([^;]*)', webpage, 'settings'), display_id)
 
         return {
-            'id': video_id,
-            'title': title,
-            'formats': formats,
+            'id': content['category']['unit']['topic']['EntryID'],
+            'display_id': display_id,
+            'title': remove_end(settings['title'], ' - BrainPOP'),
+            'description': settings['description'],
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: This should be non-fatal

File: youtube_dl/extractor/brainpop.py [hunk 11]
```diff
@@ -2,42 +2,44 @@
...
+        global_content = self._parse_json(self._html_search_regex(r'var global_content = ([^;]*)', webpage, 'global content').replace("'", '"'), display_id)
+        cdn_path = global_content.get('cdn_path', 'https://cdn.brainpop.com')
+        movie_cdn_path = global_content.get('movie_cdn_path', 'https://svideos.brainpop.com')
         ec_token = self._html_search_regex(r"ec_token : '([^']*)'", webpage, 'token')
 
-        settings = self._parse_json(self._html_search_regex(r'var settings = ([^;]*)', webpage, 'settings'), video_id)
-        title = settings['title']
-        description = settings['description']
+        screenshots = content['category']['unit']['topic'].get('screenshots', {})
+        thumbnails = [{'url': cdn_path + screenshot} for screenshot in screenshots]
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: You can cut this down to one statement

File: youtube_dl/extractor/brainpop.py [hunk 12]
```diff
@@ -2,42 +2,44 @@
...
     }
 
     def _real_extract(self, url):
-        video_id = self._match_id(url)
-        webpage = self._download_webpage(url, video_id)
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
 
-        self.report_extraction(video_id)
+        content = self._parse_json(self._html_search_regex(r'var content = ([^;]*)', webpage, 'content'), display_id)
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Just get `['category']['unit']['topic']` right away, so you don't need to specify it every time

File: youtube_dl/extractor/brainpop.py [hunk 13]
```diff
@@ -52,14 +52,14 @@ def _real_extract(self, url):
...
+
+        settings = self._parse_json(self._html_search_regex(r'var settings = ([^;]*)', webpage, 'settings JSON', '{}'), display_id)
 
         return {
-            'id': content['category']['unit']['topic']['EntryID'],
+            'id': topic['EntryID'],
             'display_id': display_id,
-            'title': remove_end(settings['title'], ' - BrainPOP'),
-            'description': settings['description'],
+            'title': remove_end(settings.get('title', display_id), ' - BrainPOP'),
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: The `title` _should_ be mandatory

Comment (yan12125) [inline youtube_dl/extractor/brainpop.py]: Such a pattern is OK if ```title``` is missing in ```settings```. @nehalvpatel Do you have an example?

Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: `title` can also be extracted from `<title>(.*)</title>` also. One of them should be fatal, right?

Comment (yan12125) [inline youtube_dl/extractor/brainpop.py]: Either way is fine. In general ```<title>``` is better than extracting from embedded data as the former is less likely to change. If there's an example that ```<title>``` or ```settings['title']``` is missing, a fallback should be provided, otherwise the extraction should be fatal.

File: youtube_dl/extractor/brainpop.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+            'width': 480,
+        })
+        self._sort_formats(formats)
+
+        settings = self._parse_json(self._html_search_regex(r'var settings = ([^;]*)', webpage, 'settings JSON', '{}'), display_id)
+
+        return {
+            'id': topic['EntryID'],
+            'display_id': display_id,
+            'title': remove_end(settings.get('title', display_id), ' - BrainPOP'),
```
Comment (TRox1972) [inline youtube_dl/extractor/brainpop.py]: Also extract title from `<title>(.+)</title>` in case current extraction fails

**Comment by EpicCodeWizard** [comment]
Does this thing still work?
